### PR TITLE
attempt to fix #21745 ("nixos-rebuild switch" kills contaner networking)

### LINKED ISSUE #21745: "nixos-rebuild switch" kills contaner networking
## Issue description

```nixos-rebuild switch``` often (but not always) executes ```systemctl stop network-setup``` then ```systemctl start network-setup```.

It kills ```nixos-container```'s network.

### Steps to reproduce

test.nix:
```
with import <nixpkgs/nixos/lib/testing.nix> { system = builtins.currentSystem; };

(makeTest { 

  nodes = {

    client = { lib, pkgs, ... }: {
      networking.firewall.enable = false;
      services.xserver.enable = true;

      services.xserver.displayManager.auto.enable = true;

      services.xserver.windowManager.default = "icewm";
      services.xserver.windowManager.icewm.enable = true;
      services.xserver.desktopManager.default = "none";

      networking.bridges.br0.interfaces = [ "eth1" ];
      networking.interfaces = {
        eth1.ip4 = lib.mkOverride 0 [ ];
        br0.ip4 = [{ address = "192.168.1.1"; prefixLength = 24; }];
      };

      containers.webserver = {
        autoStart = true;
        privateNetwork = true;
        hostBridge = "br0";
        localAddress = "192.168.1.222/24";
        config = { 
          networking.firewall.allowPing = true;
        };
      };

    };
  };

  testScript = ''
    startAll;

    $client->waitForX;

    $client->sleep(1);
    $client->execute("xterm &");
    $client->sleep(1);
    $client->sendChars("ping 192.168.1.222\n");

    $client->sleep(1);
    $client->execute("xterm &");
    $client->sleep(1);
    $client->sendChars("nixos-container root-login webserver\n");
    $client->sendChars("ping 192.168.1.1\n");

    $client->sleep(5);
    $client->execute("xterm &");
    $client->sleep(1);
    $client->sendChars("systemctl stop network-setup\n");
    $client->sleep(1);
    $client->sendChars("systemctl start network-setup\n");

    $client->sleep(10);
    $client->screenshot("screen");

    $client->sleep(10000); # allow interaction
  '';

}).driver
```

run as: (https://github.com/NixOS/nixpkgs/pull/21697 may be required for ```sendChars``` to send '.' and '-')
```
tests='eval $ENV{testScript}' $(nix-build test.nix)/bin/nixos-test-driver
```

screen

## Technical details

* System: (NixOS: `nixos-version`, Ubuntu/Fedora: `lsb_release -a`, ...)
17.03.git.7215118 (Gorilla)

* Nix version: (run `nix-env --version`)
nix-env (Nix) 1.11.5

* Nixpkgs version: (run `nix-instantiate --eval '<nixpkgs>' -A lib.nixpkgsVersion`)
"17.03.git.7215118c1b"

### PR DESCRIPTION
**Pull request by volth**
Do not merge it yet, just a demo for #21745, it needs feedback and improvement

### CODE REVIEW

**Comment by globin** [comment]
While at it could you write a NixOS test for this issue? :)

**Comment by volth** [comment]
@kampfschlaefer [made](https://github.com/NixOS/nixpkgs/compare/master...kampfschlaefer:feature/21745-fix_networking_restart#diff-49441efef2a23e90f8b1d5ad628d51a2) the headless test. Not in PR yet. And we have not yet decided how to fix the issue, there are 3 (or already 4?) ways to do it and two proposed patches.

File: nixos/modules/tasks/network-interfaces-scripted.nix [hunk 1]
```diff
@@ -180,9 +180,9 @@ in
           { description = "Virtual Network Interface ${i.name}";
             bindsTo = [ "dev-net-tun.device" ];
             after = [ "dev-net-tun.device" "network-pre.target" ];
-            wantedBy = [ "network-setup.service" (subsystemDevice i.name) ];
+            wantedBy = [ "network-setup.service" "network-addresses-${i.name}.service" "network-link-${i.name}.service" (subsystemDevice i.name) ];
```
Comment (fpletz) [inline nixos/modules/tasks/network-interfaces-scripted.nix]: Can you explain what problem you solve by adding these units to `wantedBy` and `before`?

Comment (volth) [inline nixos/modules/tasks/network-interfaces-scripted.nix]: https://github.com/NixOS/nixpkgs/issues/21745#issuecomment-278256823

**Comment by kampfschlaefer** [comment]
I strongly advise against these additional dependencies. A imho much better approach is to make the bridge service reload on change. I am working on that but haven't yet had time to fix the issue of removing interfaces that where previously added via nix-config but are now removed. Adding new devices, changing (r)stp and preserving container devices in the bridge already works with the test mentioned above.

**Comment by volth** [comment]
The dependencies target another bug: posibility of concurrent execution of ```xxx-netdev```` and ```network-addresses-xxx``` and forgetting IP configuration after network restart (not forgetting of enslaved interfaces). The bridge bug revealed it and at the moment I dont have another testcase which can show the bug besides the testcase of the bridge bug. But I believe if it is possible here that ```network-adresses-br0``` starts working before ```br0-netdev``` set the interface up, it could be possible with other interfaces as well. So I added it here "just in case" to increase reliability, instead of waiting until the problem reappears, until a bridge or another interface forgets its IP after ```nixos-rebuild```

**Comment by fpletz** [comment]
Thanks for your contribution! After careful testing in the last few days, I think we will go with #22850.

**Pull request closed without merging**